source/common/stats/thread_local_store.cc

```diff
@@ -329,12 +328,62 @@ StatType& ThreadLocalStoreImpl::ScopeImpl::safeMakeStat(
   // If we have a TLS cache, insert the stat.
   if (tls_cache) {
     tls_cache->insert(std::make_pair((*central_ref)->statName(), *central_ref));
+    ENVOY_LOG_MISC(debug, "stat added to central cache map");
+    name.debugPrint();
+    ENVOY_LOG_MISC(debug, "central_cache_map = {}", static_cast<void*>(&central_cache_map));
+    ENVOY_LOG_MISC(debug, "this = {}", static_cast<const void*>(this));
+    ENVOY_LOG_MISC(debug, "parent_ = {}", static_cast<void*>(&parent_));
   }
 
   // Finally we return the reference.
   return **central_ref;
 }
 
+// TODO(ahedberg): This is largely duplicated from safeMakeStat and should
+// be cleaned up at some point.
+template <class StatType>
+const StatType* ThreadLocalStoreImpl::ScopeImpl::safeGetStat(
+    StatName name, StatMap<std::shared_ptr<StatType>>& central_cache_map,
+    StatMap<std::shared_ptr<StatType>>* tls_cache, StatNameHashSet* tls_rejected_stats) const {
+
+  if (tls_rejected_stats != nullptr &&
+      tls_rejected_stats->find(name) != tls_rejected_stats->end()) {
+    return nullptr;
+  }
+
+  // If we have a valid cache entry, return it.
+  if (tls_cache) {
```

Comment:
I wonder if we should skip the tls cache short-circuit if we are holding the lock anyway. It would simplify this code, and we might as well just do one lookup.

When I suggested this overall strategy I had forgotten you'd need to hold the lock while looping over scopes. Sorry about that again.

Reply:
Done.